SUS-5855 | use Redis as a temporary metrics storage for Prometheus

macbre opened the pull request:
https://wikia-inc.atlassian.net/browse/SUS-5855

Prometheus' Pushgateway does not behave they way we'd like it to. Let's use Redis as a temporary cross-server (and cross-pods) storage and expose `/metrics.php` HTTP end-point to be used by Prometheus.

`Wikia\Metrics\Collector` API did not change.

### An example

```
$ curl -s 'http://community.macbre.wikia-dev.pl/metrics.php'
# HELP mediawiki_celery_tasks_delay_seconds How long task has waited in the queue before being executed
# TYPE mediawiki_celery_tasks_delay_seconds gauge
mediawiki_celery_tasks_delay_seconds{task_call="Wikia\\Tasks\\Tasks\\MultiLookupTask::updateMultiLookup",task_status="completed",env="dev"} 954500.01632905
# HELP mediawiki_celery_tasks_duration_seconds Time it took to execute the task
# TYPE mediawiki_celery_tasks_duration_seconds gauge
mediawiki_celery_tasks_duration_seconds{task_call="Wikia\\Tasks\\Tasks\\MultiLookupTask::updateMultiLookup",task_status="completed",env="dev"} 0.54292297363281
# HELP mediawiki_celery_tasks_total Number of Celery tasks executed
# TYPE mediawiki_celery_tasks_total counter
mediawiki_celery_tasks_total{task_call="Wikia\\Tasks\\Tasks\\MultiLookupTask::updateMultiLookup",task_status="completed",env="dev"} 6
```

**But**, before we can merge it we need to have `redis` pecl package installed across all servers and in Docker container.

macbre committed:
SUS-5855 | use Redis as a temporary metrics storage for Prometheus

macbre committed:
Docker | expose /metrics.php

macbre committed:
Collector | handle Redis exceptions

macbre committed:
Merge branch 'dev' into SUS-5855-prometheus-redis

macbre committed:
Merge remote-tracking branch 'origin/dev' into SUS-5855-prometheus-redis

macbre committed:
metrics.php - no caching

macbre committed:
SUS-5855 | collect Prometheus metrics from /metrics.php app end-point

macbre committed:
Collect /metrics.php from mediawiki-tasks as well

pwojcik86 reviewed (commented):
I think we can simplify a bit if we add another label to `mediawiki-prod` and `mediawiki-tasks`. Then we can have one `servicemonitor` looking at all pods. Something like `type: mediawiki`.

macbre committed:
metrics.php: gather metrics from mediawiki and mediawiki-tasks deployments

macbre committed:
And sandboxes

macbre committed:
Labels for ServiceMonitor for /metrics.php

hakubo commented on metrics.php:
```diff
@@ -0,0 +1,21 @@
+$registry = new CollectorRegistry($adapter);
+$renderer = new RenderTextFormat();
+
+header('Content-Type: ' . RenderTextFormat::MIME_TYPE);
```
missing spaces

macbre committed:
docker/sandbox/README.md: how to get logs

macbre committed:
Merge remote-tracking branch 'origin/dev' into SUS-5855-prometheus-redis

macbre committed:
WikiFactoryLoader: emit exception with HTTP 400

$ curl -sv -H 'Host: 0' 0 0.0.0.0/metrics.php
* Rebuilt URL to: 0/
*   Trying 0.0.0.0...
* Connected to 0 (127.0.0.1) port 80 (#0)
> GET / HTTP/1.1
> Host: 0
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 400 Bad Request
< Server: nginx/1.14.0
< Date: Wed, 10 Oct 2018 07:40:24 GMT
< Content-Type: text/html; charset=UTF-8
< Transfer-Encoding: chunked
< Connection: keep-alive
< X-Content-Type-Options: nosniff
<
Cannot tell which wiki it is (neither SERVER_NAME, SERVER_ID nor SERVER_DBNAME is defined)
* Connection #0 to host 0 left intact

macbre committed:
/metrics.php: tell WikiFactoryLoader class which wiki to use when serving this request
